text/cognito-construct-library/working-backwards-readme.md

```diff
@@ -0,0 +1,300 @@
+
+*Note*: By default, the email and/or phone number attributes will be marked as required if they are specified as a
+verified contact method. See [attributes](#attributes) section for details about standard attributes.
+
+> TODO: pre-signup lambda trigger
+
+The `verifyContactMethods` attribute allows for this to be configured.
+
+*Defaults*:
+* signUp.selfSignUp: true
+* signUp.tempPasswordValidity: 7 days
+* signUp.welcomeEmail.subject: 'Thanks for signing up'
+* signUp.welcomeEmail.body - 'Hello {username}, Your temporary password is {####}'
+* signUp.welcomeSms.message - 'Your temporary password is {####}'
+* signUp.verifyContactMethods - Email
+
+Code sample:
+
+```ts
+new UserPool(this, 'myuserpool', {
+  // ...
+  // ...
+  signUp: {
+    selfSignUp: true,
+    tempPasswordValidity: Duration.days(3),
+    welcomeEmail: {
+      subject: 'Welcome to our awesome app!'
+      body: 'Hello {username}, Thanks for signing up to our awesome app! Your temporary password is {####}'
+    },
+    welcomeSms: {
+      message: 'Your temporary password for our awesome app is {####}'
+    },
+    verifyContactMethods: [ ContactMethods.EMAIL, ContactMethods.PHONE_NUMBER ],
+  }
+});
+```
+
+> Internal Note: Implemented via UserPool-AdminCreateUserConfig, temp password via UserPool-Policies,
+> verifyContactMethods via UserPool-AutoVerifiedAttributes.
+
+### Sign In
+
+These are the various ways a user of your app can sign in. There are 4 options available with the enum `SignInType`:
+
+* USERNAME: Allow signing in using the one time immutable user name that the user chose at the time of sign up.
+* PREFERRED\_USERNAME: Allow signing in with an alternate user name that the user can change at any time. However, this
+  is not available if the USERNAME option is not chosen.
+* EMAIL: Allow signing in using the email address that is associated with the account.
+* PHONE\_NUMBER: Allow signing in using the phone number that is associated with the account.
+
+*Defaults*: USERNAME.
+
+Code sample:
+
+```ts
+new UserPool(this, 'myuserpool', {
+  // ...
+  // ...
+  signInType: [ SignInType.USERNAME, SignInType.EMAIL ],
+});
+```
+
+> Internal Note: Implemented via UserPool-UsernameAttributes and -AliasAttributes
+
+### Attributes
+
+These are the set of attributes you want to collect and store with each user in your user pool. Cognito provides a set
+of standard attributes that are available all user pools. Users are allowed to select any of these standard attributes
+to be required. Users will not be able to sign up without specifying the attributes that are marked as required. Besides
+these, additional attributes can be further defined, known as custom attributes.
+
+Custom attributes cannot be marked as required.
+
+[Go here](https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-settings-attributes.html) for more info.
```

Comment:
"Go here" is not an ideal link title. Something like "See [User Pool Attributes](link) for more information".